lib/child_process.js

```diff
@@ -200,7 +201,10 @@ exports.execFile = function(file /*, args, options, callback*/) {
   }
 
   if (!callback && pos < arguments.length && arguments[pos] != null) {
-    throw new errors.TypeError('ERR_INVALID_ARG_VALUE', 'args', arguments[pos]);
+    throw new errors.TypeError(
+      'ERR_INVALID_SIGNATURE',
+      'child_process.execFile', '(file[, args][, options][, callback])',
```

Comment:
What I am still missing is the type for the arguments. So we might just use a TypeScript annotation.
As in:

```js
(file: string, args?: array, options?: object, callback?: function)
// or
(file: string[, args: array][, options: object][, callback: function])
// or the way we do it in our docs:
(file <string> [, args <string[]> ][, options <Object> ][, callback <Function> ])
// In case something has more than one valid type, we just use a union type:
(file: string[, args: array | number][, options: object | string])
// or again like in our docs:
(file <string>, <Uint8Array> [, args <string[]>, <string>])
```

Reply:
I don't think we should do this in the code before the documentation starts to do it...if we are not being careful this could result in disparity between the docs and the error messages. Also there are still opinions about TypeScript v.s. Flow annotations, so I would like to see those get sorted for the docs first.

Reply:
The way this error message works will always be work to keep up to date with the documentation. And that is not that easy. As soon as we add a single way of dealing with something, it would have to be updated again, even though it might not have touched that part at all.

Without the types, I still feel the errors are not expressive enough and could create quite some confusion instead.

I personally think we should use the way the use it in our docs with e.g. `file <string>` but I would be fine with any way of showing the type.

Reply:
I agree without the types the errors are not expressive enough, but for people who are not familiar with the TypeScript annotations, those kinds of signatures might be even more confusing than our optional syntax..

Reply:
My preference would be:

1. Update the docs to put types in the signature
2. Update the errors and copy the signatures verbatim

The current optional syntax is already quite confusing, but at least that's what people who read the docs are familiar with. I don't think this PR should invent another signature syntax with types (the current one is somewhat homegrown already and that's why it's sometimes confusing to newcomers), or use a existing one because this is always opinionated and requires another round of discussion. I remember there is already an issue asking the docs to change the homegrown optional syntax into something a type-checker can check.

Reply:
I guess the most intuitive is the one we use in our docs? Or what do you think is best to solve this?

Reply:
Well, the users can always look at the docs for the types. I just don't want to invent a signature with types or pick an existing one in this PR. No matter how we put the types here, it's always gonna confuse newcomers even more than the current optional syntax does - I can live with that one because that's already in the docs, but I don't think a confusing and not-well-recognized signature is going to make it more helpful, sooner or later there might be issues asking why we are trying to invent a new syntax or why we chose A over B.

Reply:
I do not worry all that much about that. Since we use our new error system, what stays is the error code. And the message can be changed at any point of time if someone feels like changing it.

Reply:
In that case I think we can just leave it here, because it is just copying what the doc has. The signature-with-types syntax deserve another PR to discuss. The current error requires the user to look up the docs to find out what's going on anyway.

```
> child_process.execFile('ls', {}, [1,2,3])
TypeError [ERR_INVALID_ARG_VALUE]: The argument 'args' is invalid. Received [ 1, 2, 3 ]
    at Object.exports.execFile (child_process.js:203:11)
> child_process.fork('ls', 'test')
TypeError [ERR_INVALID_ARG_VALUE]: The argument 'arguments[1]' is invalid. Received 'test'
    at Object.exports.fork (child_process.js:68:13)
```

Reply:
I am not sure I can follow. You mean you suggest to keep it as is? To me it definitely feels like this PR is a good idea but we just have to find a way to agree on one way to visualize the types.

Reply:
@BridgeAR Yes, I meant I want to keep it as is, with the signatures copied from the documentation, because I personally don't feel comfortable picking an opinionated typed signature syntax myself - if I pick one and there are people asking me why, I am afraid I cannot come up with an answer because given how complicated our overloads are, I could not see anyway we can squeeze the type information into one line with the optional arguments syntax and still be intuitive and readable instead of making things even harder to understand.

What I might be comfortable with though, is listing the type information in line-separated bullet points, as in the documentation, but then...why not just read the documentation? The point of these error codes is that they'll be easier to look up in the documentation.